internal/pkg/term/syncbuffer/syncbuffer.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a goroutine safe bytes.Buffer as well printing functionality to the terminal.
+package syncbuffer
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+)
+
+// fileReader interface that can read data from a file.
+// It extends the io.Reader interface, which provides a method for reading bytes into a buffer.
+type fileReader interface {
+	io.Reader
+}
+
+// SyncBuffer is a synchronized buffer that can be used to store output data and coordinate between multiple goroutines.
+type SyncBuffer struct {
+	bufMu sync.Mutex    // bufMu is a mutex that can be used to protect access to the buffer.
+	buf   bytes.Buffer  // buf is the buffer that stores the data.
+	done  chan struct{} // done is a channel that can be used to signal when the operations are complete.
+}
+
+// NewSyncBuffer creates and returns a new SyncBuffer object with an initialized 'done' channel.
+func NewSyncBuffer() *SyncBuffer {
+	return &SyncBuffer{
+		done: make(chan struct{}),
+	}
+}
+
+// Write appends the given bytes to the buffer.
+func (b *SyncBuffer) Write(p []byte) (n int, err error) {
+	b.bufMu.Lock()
+	defer b.bufMu.Unlock()
+	return b.buf.Write(p)
+}
+
+// strings returns an empty slice if the buffer is empty.
+// Otherwise, it returns a slice of all the lines stored in the buffer.
+func (b *SyncBuffer) strings() []string {
+	b.bufMu.Lock()
+	defer b.bufMu.Unlock()
+	lines := b.buf.String()
+	if len(lines) == 0 {
+		return nil
+	}
+	return strings.Split(strings.TrimSpace(lines), "\n")
+}
+
+// IsDone returns true if the Done channel has been closed, otherwise return false.
+func (b *SyncBuffer) IsDone() bool {
+	select {
+	case <-b.done:
+		return true
+	default:
+		return false
+	}
+}
+
+// MarkDone closes the Done channel.
+func (b *SyncBuffer) MarkDone() {
+	close(b.done)
+}
+
+// CopyToBuffer reads from a fileReader and writes the data to the buffer.
+func (b *SyncBuffer) CopyToBuffer(r fileReader) error {
+	return copyFileToBuffer(b, r)
+}
+
+// LabeledSyncBuffer is a struct that combines a SyncBuffer with a string label.
+type LabeledSyncBuffer struct {
+	label   string
+	syncBuf *SyncBuffer
+}
+
+// New creates and returns a new LabeledSyncBuffer without label.
+func New(syncBuf *SyncBuffer) *LabeledSyncBuffer {
+	return &LabeledSyncBuffer{
+		syncBuf: syncBuf,
+	}
+}
+
+// NewWithLabel creates and returns a new LabeledSyncBuffer with the given label and SyncBuffer.
+func NewWithLabel(label string, syncBuf *SyncBuffer) *LabeledSyncBuffer {
+	return &LabeledSyncBuffer{
+		label:   label,
+		syncBuf: syncBuf,
+	}
+}
+
+// CopyToLabeledBuffer reads from a fileReader and writes the data to the buffer of LabeledSyncBuffer.
+func (b *LabeledSyncBuffer) CopyToLabeledBuffer(r fileReader) error {
```

Comment:
I wonder if we need this?
I'm thinking clients can use the following flow:
```go
buf, err := syncbuffer.Copy(someReader /* such as a pipe reader */)
if err != nil { return err }
labeledBuf := buf.WithLabel("title")
```

Reply:
🤔 My major concern with Copy returning *SyncBuffer `func Copy(r io.Reader) (*SyncBuffer, error)`
 as constructor is that we need to wait until all the data is copied to the buffer before printing from it.

💭 The desired approach is to print logs from the buffer as soon as we get some data in it. I.e Both copying and printing should go in parallel.   
```
lbSyncBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
count := 0 

g.Go(func() error {
  syncBuf, err := syncbuffer.Copy(pr) // Wait until the copy to the buffer is done and return the SyncBuffer, then attach the label
  lbSyncBuf := syncBuf.WithLabel(label)
  mu.Lock()
  lbSyncBuffers[count] = lbSyncBuf
  count++
  defer mu.Unlock()
  return nil
})

 ltp, err := syncbuffer.NewLabeledTermPrinter(os.Stderr, lbSyncBuffers)
 
 g.Go(func() error {
        // Print Logic
})
 ```
👀 However, when creating the `LabeledTermPrinter` object, the `lbSyncBuffers` will be nil because the copy goroutine has not completed the assigned work yet.
This means that by the time the Print goroutine runs, we do not have any data in the buffers of LabeledTermPrinter.


 To address this issue, What I want is to create a new SyncBuffer directly and then attaching a label to it instead of copy returning the SyncBuffer. This way, we can print logs as soon as we get the data into the buffer:

```
// An array of LabeledSyncBuffers
lbSyncBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
count := 0 

// Create a new SyncBuffer and attach a label to it
syncBuf := syncbuffer.New()
lbSyncBuf := syncBuf.WithLabel(label)
lbSyncBuffers[count] = lbSyncBuf
count++

g.Go(func() error {
if _, err := syncbuffer.Copy(syncBuf,pr)
})

 ltp, err := syncbuffer.NewLabeledTermPrinter(os.Stderr, lbSyncBuffers)

g.Go(func() error {
// Print Logic
})
```
🤔 I am going to modify the Copy function as below if the reasoning sounds good:
`func Copy( syncBuf *SyncBuffer, r io.Reader) (error)`